[android][fix] Tensor class created from java does not call native methods

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31520 [android][fix] Tensor class created from java does not call native methods**

Differential Revision: [D19199477](https://our.internmc.facebook.com/intern/diff/D19199477)
Fix for the issue: https://github.com/pytorch/pytorch/issues/31419

After this fix `org.pytorch.Tensor` if created from java will not create it's cxx part (at::Tensor) - it will be converted to it lazily during `Module.forward/runMethod` as it was before making Tensor hybrid in https://github.com/pytorch/pytorch/commit/930d0751e6a3006c6251c095ead4dcab71b83d52

Test Plan:
Changes in test_app/MainActivity.java to call Tensor.fromBlob before Module.load
```
      mInputTensor = Tensor.fromBlob(mInputTensorBuffer, new long[]{1, 3, 224, 224});
      mModule = PyTorchAndroid.loadModuleFromAsset(getAssets(), BuildConfig.MODULE_ASSET_NAME);
```
gradle -p android test_app:installMobNet2QuantDebug

Before change - fails, Unsatisfied link error
After - works ok

IvanKobzarev committed:
[android][fix] Tensor class needs to load jni lib

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 25171cc:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=25171ccbb8c21a3c2ffb913edea0107d49964052).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_windows_test2</sup> | <sup>Checkout code</sup> | New in PR |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

IvanKobzarev committed:
Update on "[android][fix] Tensor class needs to load jni lib"


Differential Revision: [D19199477](https://our.internmc.facebook.com/intern/diff/D19199477)
Fix for the issue: https://github.com/pytorch/pytorch/issues/31419
After java `org.pytorch.Tensor`  become a hybrid - it needs to have jni lib loaded as it has native method `initHybrid`

IvanKobzarev committed:
Update on "[android][fix] Tensor class created from java does not call native methods"


Differential Revision: [D19199477](https://our.internmc.facebook.com/intern/diff/D19199477)
Fix for the issue: https://github.com/pytorch/pytorch/issues/31419

After this fix `org.pytorch.Tensor` if created from java will not create it's cxx part (at::Tensor) - it will be converted to it lazily during `Module.forward/runMethod` as it was before making Tensor hybrid in https://github.com/pytorch/pytorch/commit/930d0751e6a3006c6251c095ead4dcab71b83d52

Test Plan:
Changes in test_app/MainActivity.java to call Tensor.fromBlob before Module.load
```
      mInputTensor = Tensor.fromBlob(mInputTensorBuffer, new long[]{1, 3, 224, 224});
      mModule = PyTorchAndroid.loadModuleFromAsset(getAssets(), BuildConfig.MODULE_ASSET_NAME);
```
gradle -p android test_app:installMobNet2QuantDebug

Before change - fails, Unsatisfied link error
After - works ok

IvanKobzarev committed:
Update on "[android][fix] Tensor class created from java does not call native methods"


Differential Revision: [D19199477](https://our.internmc.facebook.com/intern/diff/D19199477)
Fix for the issue: https://github.com/pytorch/pytorch/issues/31419

After this fix `org.pytorch.Tensor` if created from java will not create it's cxx part (at::Tensor) - it will be converted to it lazily during `Module.forward/runMethod` as it was before making Tensor hybrid in https://github.com/pytorch/pytorch/commit/930d0751e6a3006c6251c095ead4dcab71b83d52

Test Plan:
Changes in test_app/MainActivity.java to call Tensor.fromBlob before Module.load
```
      mInputTensor = Tensor.fromBlob(mInputTensorBuffer, new long[]{1, 3, 224, 224});
      mModule = PyTorchAndroid.loadModuleFromAsset(getAssets(), BuildConfig.MODULE_ASSET_NAME);
```
gradle -p android test_app:installMobNet2QuantDebug

Before change - fails, Unsatisfied link error
After - works ok
